### PR TITLE
Filters for cognitive search infra.

### PR DESCRIPTION
**Pull request by spanersoraferty**
Apologies this is a bit of a big lump but most of this stuff has been carried over from work we did on the CSCP re-write so hopefully is familiar to @CathLass. I've added extensive code comments and tested the results through a plugin which for the most part look ok but feel free to challenge anything.

### CODE REVIEW

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/SearchFilterExpressionBuilderTests.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+using Dfe.Data.Common.Infrastructure.CognitiveSearch.Tests.Filtering.FilterExpressions.StubBuilders;
+using Dfe.Data.Common.Infrastructure.CognitiveSearch.Tests.Filtering.TestDoubles.StubBuilders;
+using Dfe.Data.Common.Infrastructure.CognitiveSearch.Tests.Search.TestDoubles;
+using FluentAssertions;
+using Microsoft.Extensions.Options;
+using Xunit;
+
+namespace Dfe.Data.Common.Infrastructure.CognitiveSearch.Tests.Filtering;
+
+public class SearchFilterExpressionBuilderTests
```
Comment (CathLass) [inline Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/SearchFilterExpressionBuilderTests.cs:15]: Only happy path tests

**Review by CathLass — COMMENTED** [review]
ReadMe hasn't been updated with info about filtering. Might be a case for cutting down the details in the ReadMe just to make it less work to keep it updated.

A bunch of comments - as usual up for discussion, I've made no value judgements about which ones are important.

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/FilterExpressions/SearchInFilterExpression.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
+﻿using Dfe.Data.Common.Infrastructure.CognitiveSearch.Filtering.FilterExpressions.Formatters;
+
+namespace Dfe.Data.Common.Infrastructure.CognitiveSearch.Filtering.FilterExpressions;
+
+/// <summary>
+/// Creates an OData filter expression which represents the <b>search.in</b> function supported by
+/// Azure AI search filter capability. When using search.in the values/variables provisioned
+/// are generally prescribed to string formats (it cannot handle boolean types for example).
```
[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/FilterExpressions/SearchInFilterExpression.cs:8]: Yeah, this is logic you put in originally

[the remark this answers is not in the document]
Comment (CathLass) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/FilterExpressions/SearchInFilterExpression.cs:8]: OK, so was I wrong then or am I wrong now?

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/FilterExpressions/Factories/SearchFilterExpressionFactoryTests.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        ISearchFilterExpression searchFilterExpression =
+            searchFilterExpressionFactory.CreateFilter(typeof(SearchInFilterExpression));
+
+        // assert.
+        searchFilterExpression.Should().NotBeNull();
+        searchFilterExpression.Should().BeOfType<SearchInFilterExpression>();
+    }
+
+    [Fact]
+    public void CreateFilter_WithValidFilterGenricTemplateType_ReturnsExpectedConfiguredFilterFromfactory()
```
[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/FilterExpressions/Factories/SearchFilterExpressionFactoryTests.cs]: Done

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/FilterExpressions/Factories/SearchFilterExpressionFactoryTests.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        ISearchFilterExpression searchFilterExpression =
+            searchFilterExpressionFactory.CreateFilter<SearchInFilterExpression>();
+
+        // assert.
+        searchFilterExpression.Should().NotBeNull();
+        searchFilterExpression.Should().BeOfType<SearchInFilterExpression>();
+    }
+
+    [Fact]
+    public void CreateFilter_WithNullFilterName_ThrowsExpectedException()
```
[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/FilterExpressions/Factories/SearchFilterExpressionFactoryTests.cs:59]: Yeah assuming we don't permit warnings. I'll leave em in for now.

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/FilterExpressions/SearchGeoLocationFilterExpression.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        // We expect only two parameters here, representing Latitude and longitude.
+        if (searchFilterRequest.FilterValues.Length != 2){
+            throw new ArgumentException(
+                "The geo-location filter expression expects two values representing latitude and longitude.", searchFilterRequest.FilterKey);
+        }
+
+        // Ensure the geo-location points are in the correct format.
+        searchFilterRequest.FilterValues.ToList()
+            .ForEach(filterValue => {
+                if (!float.TryParse(filterValue.ToString(), out _)){
```
[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/FilterExpressions/SearchGeoLocationFilterExpression.cs:63]: Who am I to argue with Chat GPT, I'll make the change ;)

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/FilterExpressions/SearchInFilterExpression.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/FilterExpressions/SearchInFilterExpression.cs:8]: Errrr not sure, pretty sure it was causing issues at the time which is why you added it ;)

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/LogicalOperators/Factories/LogicalOperatorFactoryTests.cs [hunk 6]
```diff
@@ -0,0 +1,112 @@
+﻿using Dfe.Data.Common.Infrastructure.CognitiveSearch.Filtering.LogicalOperators;
+using Dfe.Data.Common.Infrastructure.CognitiveSearch.Filtering.LogicalOperators.Factories;
+using Dfe.Data.Common.Infrastructure.CognitiveSearch.Tests.Filtering.FilterExpressions;
+using FluentAssertions;
+using Xunit;
+
+namespace Dfe.Data.Common.Infrastructure.CognitiveSearch.Tests.Filtering.LogicalOperators.Factories;
+
+public sealed class LogicalOperatorFactoryTests
```
[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/LogicalOperators/Factories/LogicalOperatorFactoryTests.cs:9]: Damn, you spotted this. I had the same thought(s) when I was writing the test tbh :)

[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/LogicalOperators/Factories/LogicalOperatorFactoryTests.cs:9]: I will leave for now and add a ticked to do integration tests here

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/SearchFilterExpressionBuilderTests.cs (hunk 1 above)
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/SearchFilterExpressionBuilderTests.cs:15]: Yip, I'll add a ticket to add some red path tests

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/LogicalOperators/Factories/LogicalOperatorFactory.cs [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    /// Allows creation of an <see cref="ILogicalOperator"/> instance based on the type name requested.
+    /// </summary>
+    /// <param name="logicalOperatorName">
+    /// The name of the concrete implementation type <see cref="ILogicalOperator"/> requested.
+    /// </param>
+    /// <returns>
+    /// The configured instance of the <see cref="ILogicalOperator"/> type.
+    /// </returns>
+    /// <exception cref="ArgumentNullException">
+    /// Exception thrown if an invalid filter name string is provisioned.
```
[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/LogicalOperators/Factories/LogicalOperatorFactory.cs:78]: Done

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/SearchFilterRequest.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+   /// <param name="filterValues">
+   /// The collection of values to apply to the OData filter expression type.
+   /// </param>
+   /// <exception cref="ArgumentException">
+   /// The exception thrown if no filter values are provisioned.
+   /// </exception>
+    public SearchFilterRequest(string filterKey, IEnumerable<object> filterValues)
+    {
+        ArgumentException.ThrowIfNullOrWhiteSpace(filterKey);
+        ArgumentNullException.ThrowIfNull(filterValues);
```
[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/SearchFilterRequest.cs:39]: I'll add to doc.

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/FilterExpressions/Factories/SearchFilterExpressionFactoryTests.cs [hunk 9]
```diff
@@ -0,0 +1,108 @@
+﻿using Dfe.Data.Common.Infrastructure.CognitiveSearch.Filtering.FilterExpressions;
+using Dfe.Data.Common.Infrastructure.CognitiveSearch.Filtering.FilterExpressions.Factories;
+using FluentAssertions;
+using Xunit;
+
+namespace Dfe.Data.Common.Infrastructure.CognitiveSearch.Tests.Filtering.FilterExpressions.Factories;
+
+public class SearchFilterExpressionFactoryTests
```
[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Tests/Filtering/FilterExpressions/Factories/SearchFilterExpressionFactoryTests.cs:8]: I'll raise a ticket for this extra testing

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/FilterExpressions/LessThanOrEqualToExpression.cs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+        // Ensure we only receive a single filter value in the request.
+        if (searchFilterRequest.FilterValues.Length != 1){
+            throw new ArgumentException(
+                "Less than or equal to expression expects only one value.", searchFilterRequest.FilterKey);
+        }
+
+        // Ensure the less than or equal to filter values are in the correct format.
+        if (!double.TryParse(searchFilterRequest.FilterValues.Single().ToString(), out double number) || number <= 0){
+            throw new ArgumentException(
+                "Less than or equal to expression must be assigned a positive number or zero.", searchFilterRequest.FilterKey);
```
[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/FilterExpressions/LessThanOrEqualToExpression.cs:61]: I'd rather handle these earlier and get a more succinct exception.

File: Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/SearchFilterRequest.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/SearchFilterRequest.cs:39]: So I was basing this on use with the geo location distance but I'm guessing you're saying there could be other occasions where may want to apply the test against negative values, which is reasonable I guess.

[the remark this answers is not in the document]
Comment (spanersoraferty) [reply Dfe.Data.Common.Infrastructure.CognitiveSearch/Filtering/SearchFilterRequest.cs:39]: Done

**Review by CathLass — APPROVED** [review]

**Pull request merged**